qiskit/transpiler/basepasses.py

```diff
@@ -114,10 +184,70 @@ def __call__(self, circuit, property_set=None):
 class AnalysisPass(BasePass):  # pylint: disable=abstract-method
     """An analysis pass: change property set, not DAG."""
 
-    pass
+    def execute(
+        self,
+        passmanager_ir: DAGCircuit,
+        state: PassState | None = None,
+        callback: Callable = None,
+    ) -> DAGCircuit:
+        """Execute optimization task for input passmanager IR.
+
+        Args:
+            passmanager_ir: Passmanager IR to optimize.
+            state: A local state information associated with this optimization workflow.
+            callback: A callback function which is caller per execution of optimization task.
+
+        Returns:
+            Optimized passmanager IR.
+        """
+        # analysis pass must not mutate DAG object
+        fenced_dag = FencedDAGCircuit(passmanager_ir)
+
+        # analysis pass returns nothing, just mutates the property set
+        super().execute(
+            passmanager_ir=fenced_dag,
+            state=state,
+            callback=callback,
+        )
+
+        return passmanager_ir
 
 
 class TransformationPass(BasePass):  # pylint: disable=abstract-method
     """A transformation pass: change DAG, not property set."""
 
-    pass
+    def execute(
+        self,
+        passmanager_ir: DAGCircuit,
+        state: PassState | None = None,
+        callback: Callable = None,
+    ) -> DAGCircuit:
+        """Execute optimization task for input passmanager IR.
+
+        Args:
+            passmanager_ir: Passmanager IR to optimize.
+            state: A local state information associated with this optimization workflow.
+            callback: A callback function which is caller per execution of optimization task.
+
+        Returns:
+            Optimized passmanager IR.
+        """
+        new_dag = super().execute(
+            passmanager_ir=passmanager_ir,
+            state=state,
+            callback=callback,
+        )
+
+        if self.state.previous_run == RunState.SUCCESS:
+            self._state.completed_passes.intersection_update(set(self.preserves))
+
+            if isinstance(new_dag, DAGCircuit):
+                # Copy calibration data from the original program
+                new_dag.calibrations = passmanager_ir.calibrations
+            else:
+                raise TranspilerError(
+                    "Transformation passes should return a transformed dag."
+                    f"The pass {self.__class__.__name__} is returning a {type(new_dag)}"
+                )
+
```

Comment:
In general, all of this looks like it's being implemented at an incorrect layer to me.  It feels like a pass _manager_ should be responsible for tracking the validity or invalidity of previous run passes, whereas here it's being done at the pass level.

`DAGCircuit.calibrations` is difficult in general.  It feels to me like it should either be the responsibility of passes that rebuild the DAG entirely to propagate (much like they're required to propagate the metadata), or it would be something that shouldn't really be attached to a single circuit object at all, but a separate object that's more backend-oriented.  This is presumably pre-exisiting, though?

Reply:
Originally, execution and status tracking of passes are responsibility of running pass manager as you suggest. Note that flow controller pipeline consists of inner flow controllers and passes, and I decided to reimplement them with composite pattern. This drastically reduces the overhead of type checking and normalization. This is why pass execution is responsibility of pass itself. Without doing this, I need to write many branching logic to implement flow controller execution, which makes the code more complicated.

> It feels to me like it should either be the responsibility of passes that rebuild the DAG entirely to propagate

Yes, this is correct. However, `.calibrations` was introduced with pulse gate and we needed to handle this in baseclass because there were many transformer passes which were not aware of calibration attribute. 

> it would be something that shouldn't really be attached to a single circuit object at all, but a separate object that's more backend-oriented

It depends of what you want to achieve. When you add new basis gates, then calibration can go more global configuration. However, if you want to write something like Rabi experiment, that defines an opaque gate for Rabi drive with parametric amplitude, then calibration must be added to each circuit with different amplitudes.

Reply:
I guess I would have proposed a different model for managing calibrated gates on circuits personally, because I think the current form of `.calibrations` leads to this problem, but that's not important for this discussion, because regardless we need to work with the model we actually have.

I think given the constraints we're working in, what we've got here does work fine, provided we document that the `execute` method is not safe to override because there's too many internal-specific invariants a custom subclass would be required to uphold, and we'd no longer be free to change them if we allowed overriding of the method.  To me, that's another reason to relax the "single pipeline" form of `FlowController`; if `IfElseController` had to manually implement this logic because it couldn't subclass `FlowController` (due to the internal-structure enforcement), it would be a lot more error-prone to update.

Reply:
Some inline comments are added in [6211a5d](https://github.com/Qiskit/qiskit/pull/10127/commits/6211a5d2978af9091eb5f474273b5810828b6d68). I also want alternative data model for pulse gates.